Trsh module

alongd opened the pull request:
Added a `trsh` module to arc.job.
Relocated ESS and server troubleshooting methods.

Still missing tests.

xiaoruiDong commented on arc/job/trsh.py:
```diff
@@ -0,0 +1,647 @@
+                        ' nosyym'.format(type=job_type, software=software, label=label))
+            ess_trsh_methods.append('cartesian')
+            trsh_keyword = 'opt=(cartesian,nosymm)'
+        elif 'unconverged' in job_status and 'fine' not in ess_trsh_methods and not fine:
```
We may want to change it to 'Unconverged'?  as it is what we have in the previous section, like `Unconverged (G9999)`.

alongd commented on arc/job/trsh.py:
```diff
@@ -0,0 +1,647 @@
+                    reason = 'Memory: Additional memory required: {0} MW'.format(float(line.split()[-2]) / 1e6)
+            for line in lines[::-1]:
+                if 'the problem occurs' in line:
+                    reason = 'Errored: '
```
Changed. Do you think additional changes are required for the QChem block?

xiaoruiDong replied:
In the qchem section, as there is no way to break the loop, the line returned in `return 'errored: {0}; {1}'.format(reason, line)` is irrelevant to the problematic line. I am also thinking maybe line should be recorded only when the reason is not crystal clear. `reason` now is pretty informative and contains the key info/word of the line (True for gaussian, but maybe not true for other packages).

alongd commented on arc/job/job.py:
```diff
@@ -940,52 +940,13 @@ def troubleshoot_server(self):
+        if node is not None:
+            self.server_nodes.append(node)
+        # resubmit job
+        self.run()
```
I see in the deleted lines above that run() is called at the end of both OGE and Slurm blocks. Can you take a second look at your comment? Or did I miss something?

xiaoruiDong replied:
Please check line 1093 in the original code. There is a return, so I assume, it will not do then self.run(). And in current `troubleshoot_job_on_server()`, it becomes `return None` which returns `None` to `node`. Despite whether node is None or not, it will do self.run(). It is just a behavior difference, and not clear to me whether it will raise an error. We will know if we have a unittest for it.

alongd replied:
You're right, I added a fix. Thanks!

alongd commented:
Thanks for the review and for the excellent catches! I addressed the comments in fixups.

alongd commented:
@xiaoruiDong, I improved the error standardization by using a dictionary for the ESS job status with the following structure: `{'status': str, 'keywords': list, 'error': str, 'line': str}`, so now the keywords are easier to parse.
Please let me know if you prefer to review after I stash.
I also added tests, take a look and feel free to add additional ones (I'll be happy to discuss it next week)

xiaoruiDong commented:
@alongd Such error standardization is great and desirable. I will have a quick review of your PRs today and tomorrow. Thanks for the contributions!

alongd commented:
@xiaoruiDong, I force pushed and solved the conflicts.
Please let me know if it'll be more convenient for you to review after I squash the fixups.

xiaoruiDong commented:
@alongd Thanks. I think it is okay to keep these fixups for reviewing.

xiaoruiDong commented on arc/job/trsh.py:
```diff
@@ -227,20 +227,20 @@ def trsh_negative_freq(label, log_file, neg_freqs_trshed=None, job_types=None):
                                                                                       vibfreqs[largest_neg_freq_idx]))
             neg_freqs_idx = [largest_neg_freq_idx]  # indices of the negative frequencies to troubleshoot for
-        elif len(neg_freqs_idx) > 1 and vibfreqs[0] in neg_freqs_trshed:
+        elif len(neg_freqs_idx) > 1 and any([is_close(vibfreqs[0], vf) for vf in neg_freqs_trshed]):
```
`is_close` is a nice function. But since `numpy` offers a similar function `np.allclose()`. Do we want to use that instead, to reduce our maintenance load?

xiaoruiDong commented on arc/common.py:
```diff
@@ -470,3 +470,20 @@ def determine_ess(log_file):
+
+        bool: Whether `a` and `b` are close, `True` if they are.
+    """
+    return abs(a - b) <= max(rel_tol * max(abs(a), abs(b)), abs_tol)
```
See if we want to use `np.allclose`.

xiaoruiDong commented on arc/job/trsh.py:
```diff
@@ -110,12 +110,17 @@ def determine_ess_status(output_path, species_label, job_type, software=None):
                     done = True
+                    if 'opt' not in job_type:
+                        # if this is an opt job, we must also check that the max num of cycles hasn't been reached.
+                        break
```
I am not familiar with q-chem, but it seems like we will possibly get `Thank you very much for using Q-Chem` and `MAXIMUM OPTIMIZATION CYCLES REACHED` in an opt job. The problem is, as we assign `done = True` in this block and only assign `keywords=...` in the lower block. We will still have `done=True` when we reach the end of the loop. And the function will still return `'done'`. Probably, you want to add `done = False` before the line 153.

xiaoruiDong commented on arc/job/trsh.py:
```diff
@@ -345,6 +345,8 @@ def trsh_ess_job(label, level_of_theory, server, job_status, job_type, software,
     Returns:
         shift (str): The shift to use (only in Molpro).
+    Returns:
+        couldnt_trsh (bool): Whether a troubleshooting solution was found. `True` if it was.
```
If I understand correctly, `couldnt_trsh = True` means we are running out of solutions. If so, the comment is a little bit misleading. But I can be wrong.

alongd commented:
Thanks for reviewing! I agree with all the comments. I addressed all and squashed. Let me know if you have any additional comments. We should also add more tests at a later stage (let's talk offline)

xiaoruiDong commented:
@alongd Great! I am fine with the current PR and I agree with you that we need more test cases.
